### PR TITLE
fix(claim-dapp): add allowance check before approve

### PR DESCRIPTION
**Pull request by Gamaranto**
**Motivation**
This PR adds an allowance check before sending multiple approve transactions, to prevent users wasting money on useless approvals.

**Details**

To test this:
- Set up an env where you can redeem KPI options (can be local or the preview on Kovan / Mainnet)
- Approve the options tokens, then exit the redeem flow and refresh the page.
- You should now be able to redeem without approving again

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [x]  All existing tests pass
- [ ]  Untested

Signed-off-by: Gamaranto <francesco@umaproject.org>

### CODE REVIEW

File: src/hooks/useRedeem.ts [hunk 1]
```diff
@@ -21,12 +21,18 @@ export function useRedeem() {
...
-      );
-      addTransaction({ ...approveTx, label: "approve" });
-      await approveTx.wait();
+      const allowance = await kpiOptionsToken.allowance(account, emp.address);
+      const balance = await kpiOptionsToken.balanceOf(account);
+      const hasToApprove = allowance.lt(balance);
+      if (hasToApprove) {
+        const approveTx = await kpiOptionsToken.approve(
+          emp.address,
+          ethers.utils.parseEther("20000000000000000000")
```
Comment (nicholaspai) [inline src/hooks/useRedeem.ts:30]: What's this number, `20000000000000000000`? Some other allowance amount options:
- Maximum `uint`: [This value can be found here](https://github.com/UMAprotocol/protocol/blob/master/packages/common/src/Constants.js#L30) and sets the user's allowance to the maximum. This is arguably invasive as the user might unknowingly approve much more balance than they are comfortable with, but the benefit is that if the user trusts our dApp, then they never have to approve again.
- Just approve the `balance` amount. This is the "safest" for the user but they will have to send another approval before the next transfer

Comment (Gamaranto) [reply src/hooks/useRedeem.ts:30]: Great point, the number was a lazy replacement for `INFINITE_AMOUNT`, I think option 1 is better from a UX perspective, and I think its fine as the options only value after expiry is to be redeemed for UMA, but when approving something like a collateral type for minting tokens (ex umaverse LSP interface), then 2 would be best.
TLDR, going to change `200....0000` to the max `uint` and set it as a variable instead of an hardcoded string

File: src/hooks/useRedeem.ts [hunk 2]
```diff
@@ -6,6 +6,10 @@ import { useTransactions } from "./useTransactions";
 import { ethers } from "ethers";
 import { getKpiOptionsEMP, getKpiTokenContract } from "../utils";
 
+// max uint value is 2^256 - 1
+const MAX_UINT_VAL = ethers.BigNumber.from(2).pow(256).sub(1);
```
Comment (daywiss) [inline src/hooks/useRedeem.ts:10]: [`ethers.constants.MaxUint256 ⇒ BigNumber`](https://docs.ethers.io/v5/single-page/#/v5/api/utils/constants/)

**Review by daywiss — COMMENTED** [review]
i think there might be another issue with the await calls, but take another look and see if thats a problem

File: src/hooks/useRedeem.ts [hunk 3]
```diff
@@ -21,12 +25,18 @@ export function useRedeem() {
...
       const emp = getKpiOptionsEMP(signer, chainId);
 
-      const approveTx = await kpiOptionsToken.approve(
-        emp.address,
-        ethers.utils.parseEther("20000000000000000000")
-      );
-      addTransaction({ ...approveTx, label: "approve" });
-      await approveTx.wait();
+      const allowance = await kpiOptionsToken.allowance(account, emp.address);
+      const balance = await kpiOptionsToken.balanceOf(account);
```
[the remark this answers is not in the document]
Comment (daywiss) [reply src/hooks/useRedeem.ts:29]: nevermind, i see this is wrapped in a use callback

**Review by daywiss — APPROVED** [review]
my bad missed the callback line. looks good

File: src/hooks/useRedeem.ts (hunk 2 above)
Comment (Gamaranto) [reply src/hooks/useRedeem.ts]: thank you, changed!
